### ISSUE TITLE
Proxy starts to drop connections 

### ISSUE THREAD

**Issue by alexandrevilain** [linkerd/linkerd2#1803]
### What is the issue?

Today we started to encounter some errors for our services deployed in production with linkerd installed.
Every http request we made to an external service returned an empty body. The response was done by linkerd as our services was up. 

No configuration change was done on the cluster.

### How can it be reproduced?

It seems to be only on pods running for more than 10 days.

### Logs, error output, etc

```
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.12.3.181:34462} linkerd2_proxy::proxy::http::router service error: an error occurred trying to connect: No route to host (os error 113)
````

#### `linkerd check` output

```text
kubernetes-api: can initialize the client..................................[ok]
kubernetes-api: can query the Kubernetes API...............................[ok]
kubernetes-api: is running the minimum Kubernetes API version..............[ok]
linkerd-api: control plane namespace exists................................[ok]
linkerd-api: control plane pods are ready..................................[ok]
linkerd-api: can initialize the client.....................................[ok]
linkerd-api: can query the control plane API...............................[ok]
linkerd-api[kubernetes]: control plane can talk to Kubernetes..............[ok]
linkerd-api[prometheus]: control plane can talk to Prometheus..............[ok]
linkerd-version: can determine the latest version..........................[ok]
linkerd-version: cli is up-to-date.........................................[ok]
linkerd-version: control plane is up-to-date...............................[ok]
```

### Environment

- Kubernetes Version: v1.10.6-gke.2
- Cluster Environment: GKE
- Host OS: Container OS
- Linkerd version: stable-2.0.0

### Possible solution

To make my production app working again, I needed to remove linkerd :(

### Additional context

The external service is requested using it's public dns name on the 8080 port

**Comment by olix0r** [comment]
"No route to host" is a very odd error to see. This would typically indicate that the proxy is trying to connect to an IP address that the operating system will not route to.

> Every http request we made to an external service returned an empty body.

What do you mean by this? Is the service a Kubernetes Service? An [ExternalName](https://kubernetes.io/docs/concepts/services-networking/service/#externalname) Service? Or just a DNS name that resolves outside of the cluster?

**Comment by alexandrevilain** [comment]
It's a name that resolves outside of the cluster

**Comment by olix0r** [comment]
@alexandrevilain Do you think it's possible that the IP address of the destination changed? Perhaps the proxy had incorrectly cached an older IP address that was invalidated.

**Comment by alexandrevilain** [comment]
@olix0r
No, it’s impossible, the destination use a dedicated IP which can’t be changed.

**Comment by sokoow** [comment]
I got exactly same problem: brand new kube cluster, no RBAC or pod network policies, with canal ovaerlay. Linkerd both on tap and controller throws me 107 constantly, first you see errors connecting to pod addresses, then to nodes itself. Here's a sidecar proxy log:

```
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.42.6.63:51830} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=in listen=0.0.0.0:4143 remote=192.168.3.163:42918} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=in listen=0.0.0.0:4143 remote=192.168.3.163:42998} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=in listen=0.0.0.0:4143 remote=192.168.3.163:43144} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=in listen=0.0.0.0:4143 remote=192.168.3.163:43236} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=in listen=0.0.0.0:4143 remote=192.168.3.163:43370} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=in listen=0.0.0.0:4143 remote=192.168.3.163:43458} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.42.6.63:52536} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=in listen=0.0.0.0:4143 remote=192.168.3.163:43632} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
```

that 192.168.... address is a physical node address

here's corresponding tap log:

```
time="2018-10-24T12:07:25Z" level=info msg="Tapping 1 pods for target: {Namespace:default Type:deployment Name:webui-gateway XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}"
time="2018-10-24T12:07:25Z" level=info msg="Establishing tap on 10.42.7.84:4190"
time="2018-10-24T12:07:35Z" level=error msg="[10.42.7.84] encountered an error: rpc error: code = Canceled desc = context canceled"
```

here's linkerd-proxy log:

```
INFO linkerd2_proxy using controller at Some(HostAndPort { host: DnsName(DnsName(DNSName("localhost."))), port: 8086 })
INFO linkerd2_proxy routing on V4(127.0.0.1:4140)
INFO linkerd2_proxy proxying on V4(0.0.0.0:4143) to None
INFO linkerd2_proxy serving Prometheus metrics on V4(0.0.0.0:4191)
INFO linkerd2_proxy protocol detection disabled for inbound ports {25, 3306}
INFO linkerd2_proxy protocol detection disabled for outbound ports {25, 3306}
INFO admin={bg=resolver} linkerd2_proxy::transport::connect DNS resolved DnsName(DnsName(DNSName("localhost."))) to 127.0.0.1
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.42.7.80:46404} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.42.7.80:46400} linkerd2_proxy::proxy::tcp tcp duplex error: Connection reset by peer (os error 104)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.42.7.80:46408} linkerd2_proxy::proxy::tcp tcp duplex error: Broken pipe (os error 32)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.42.7.80:46396} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.42.7.80:46412} linkerd2_proxy::proxy::tcp tcp duplex error: Connection reset by peer (os error 104)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.42.7.80:46416} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.42.7.80:46428} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.42.7.80:46424} linkerd2_proxy::proxy::tcp tcp duplex error: Transport endpoint is not connected (os error 107)
ERR! proxy={server=out listen=127.0.0.1:4140 remote=10.42.7.80:46446} linkerd2_proxy::proxy::tcp tcp duplex error: Connection reset by peer (os error 104)
```

looks like the whole setup is a bit screwed, funny enough, web ui works quite well, except it doesn't show any data :)

**Comment by klingerf** [comment]
@alexandrevilain Are you still seeing this issue with the latest [Linkerd 2.1 release](https://github.com/linkerd/linkerd2/releases/tag/stable-2.1.0)? There have been a lot of stability fixes in the latest release. This also seems related to #1778, and I'm asking that report to verify with Linkerd 2.1 as well.

**Comment by alexandrevilain** [comment]
Hi @klingerf 

I can't test on my staging cluster. I've got the same issue as https://github.com/linkerd/linkerd2/issues/2012 .

**Comment by klingerf** [comment]
@alexandrevilain Hmm, ok -- thanks for the update. There seems to be a lot of activity on #2012, so hopefully that will get fixed soon and you can get back to testing this one.

**Comment by hawkw** [comment]
Hi @alexandrevilain, #2012 has been fixed in the latest edge release (`edge-19.1.1`) and I'd be curious to know if you're still seeing this issue on that version?
